components/backup/src/endpoint.rs

```diff
@@ -177,100 +223,113 @@ impl<E: Engine, R: RegionInfoProvider> Endpoint<E, R> {
+            region_info,
+            workers,
+            db,
+        }
+    }
+
+    fn spawn_backup_worker(
         &self,
-        brange: BackupRange,
+        prs: Arc<Mutex<Progress<R>>>,
         start_ts: u64,
         end_ts: u64,
         storage: LimitedStorage,
         tx: mpsc::Sender<(BackupRange, Result<BackupRes>)>,
+        cancel: Arc<AtomicBool>,
     ) {
         // TODO: support incremental backup
         let _ = start_ts;
 
         let backup_ts = end_ts;
-        let mut ctx = Context::new();
-        ctx.set_region_id(brange.region.get_id());
-        ctx.set_region_epoch(brange.region.get_region_epoch().to_owned());
-        ctx.set_peer(brange.leader.clone());
-        // TODO: make it async.
-        let snapshot = match self.engine.snapshot(&ctx) {
-            Ok(s) => s,
-            Err(e) => {
-                error!("backup snapshot failed"; "error" => ?e);
-                return tx.send((brange, Err(e.into()))).unwrap();
-            }
-        };
+        let engine = self.engine.clone();
         let db = self.db.clone();
         let store_id = self.store_id;
-        self.workers.spawn(lazy(move || {
-            if brange.has_canceled() {
-                warn!("backup task has canceled"; "range" => ?brange);
+        // TODO: make it async.
+        self.workers.spawn(lazy(move || loop {
+            let branges = prs.lock().unwrap().forward(WORKER_TAKE_RANGE);
+            if branges.is_empty() {
                 return Ok(());
             }
-            let snap_store = SnapshotStore::new(
-                snapshot,
-                backup_ts,
-                IsolationLevel::Si,
-                false, /* fill_cache */
-            );
-            let start_key = brange.start_key.clone();
-            let end_key = brange.end_key.clone();
-            let mut scanner = snap_store
-                .entry_scanner(start_key.clone(), end_key.clone())
-                .unwrap();
-            let mut batch = EntryBatch::with_capacity(1024);
-            let name = backup_file_name(store_id, &brange.region);
-            let mut writer = match BackupWriter::new(db, &name, storage.limiter) {
-                Ok(w) => w,
-                Err(e) => {
-                    error!("backup writer failed"; "error" => ?e);
-                    return tx.send((brange, Err(e))).map_err(|_| ());
+            for brange in branges {
+                if cancel.load(Ordering::SeqCst) {
+                    warn!("backup task has canceled"; "range" => ?brange);
+                    return Ok(());
                 }
-            };
-            let start = Instant::now();
-            loop {
-                if let Err(e) = scanner.scan_entries(&mut batch) {
-                    error!("backup scan entries failed"; "error" => ?e);
-                    return tx.send((brange, Err(e.into()))).map_err(|_| ());
+                let mut ctx = Context::new();
+                ctx.set_region_id(brange.region.get_id());
```

Comment:
The context is constructed from brange, how about moving L300-L329 to `BackupRange::backup`? 

Also, we can call `BackupWriter::save` outside of `BackupRange::backup`.

Reply:
Yes. But L322-L329 may not need move to `BackupRange::backup` if we will call `BackupWriter::save` outside of `BackupRange::backup`